### PR TITLE
feat(web): add find in conversation

### PR DESCRIPTION
**Pull request by Lucenx9**
## Problem

Native browser find only sees the part of a conversation currently mounted in the DOM. Long T3 Code threads are paginated and virtualized, so users could not reliably find text outside the visible window.

## Solution

Adds in-conversation find to the web and desktop chat surfaces with `Cmd+F` on macOS and `Ctrl+F` elsewhere, plus the same action in the command palette.

The server searches rendered user text and canonical final assistant text through a bounded, capability-gated RPC. The client gets an exact thread-wide count, loads older history when needed, expands folded content, and highlights the active result without hydrating the whole thread.

## How It Works

```mermaid
sequenceDiagram
    actor User
    participant Client as Web / Desktop
    participant Server as T3 server
    participant Projection as SQLite projection
    User->>Client: Cmd/Ctrl+F and query
    Client->>Server: findThread(query, result page)
    Server->>Projection: Read canonical text for one thread
    Projection-->>Server: Ordered message rows
    Server-->>Client: Exact count and bounded locators
    Client-->>User: Load, scroll, and highlight active match
```

### Before

The conversation has no thread-aware find control.

Before: conversation without thread-aware find

### After: active result

The find bar reports the exact count and highlights the selected occurrence.

After: find bar showing the third highlighted result

### After: unloaded history

The find bar can navigate to a result outside the initially loaded history.

After: server-backed navigation to the oldest unloaded match

### After: folded content

Folded details open automatically when they contain the active result.

After: exact highlight inside automatically expanded details

## Verification

- 501 focused tests passed in the final post-merge run across shared search logic, file-link rendering parity, projection querying, RPC transport, timeline behavior, and DOM highlighting.
- Typechecks passed for contracts, shared, client runtime, web, and server.
- Interactive web verification covered the shortcut, command-palette focus, unloaded history, next/previous navigation, folded details, no-result state, and Escape cleanup with zero browser errors.
- Two Daybreak Blue xhigh review passes finished clean after the final fixes.
- Fallow reviewed the final diff with one worker: all four structural decisions were accepted, with no stale or rejected judgments.

## Checklist

- [x] This PR addresses one focused concern
- [x] I explained what changed and why
- [x] I included before/after screenshots
- [x] No animation or timing behavior requires a video

Built with GPT-5.6-Sol via the Codex harness in T3 Code.

---

> [!NOTE]
> **Medium Risk**
> Touches projection revision semantics, thread-detail paging boundaries, and a new orchestration read RPC, but capability-gated and covered by extensive tests; main risk is edge cases in windowed history and highlight positioning on large threads.
> 
> **Overview**
> Adds **find in conversation** for paginated, virtualized threads: clients call a new `orchestration.findThread` RPC (read scope, advertised via `threadFind: 1` in server config) instead of relying on browser find over only mounted DOM.
> 
> **Server** loads searchable user and canonical assistant text from projections, returns exact totals with paginated locators and opaque `targetCursor` values. `projection_threads.searchable_messages_revision` (migration 048) invalidates cached find text when messages change, reverts run, or the canonical assistant message changes; thread-detail keyset paging gains inclusive boundaries and `targetAnchorAt` so matches can load a tight window (including turnless messages before the first turn).
> 
> **Web** adds `ThreadFindBar` (debounced query, prev/next, Cmd/Ctrl+F and command palette), `useThreadFindTarget` / `ThreadFindSource` for scrolling, loading older or direct windows via `requestThreadTurnsAround`, expanding folded markdown, and CSS Custom Highlight API highlighting (including diff shadow roots). Citation navigation shares the extended pagination model (`hasMore`, `onLoadWindow`).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4bd22c3f3c7c712e55902582f37af7da836ac0c5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add `thread.find` in-conversation search across web, server, and shared packages
> - Adds a full-stack find-in-conversation feature: a `ThreadFindBar` UI, an `orchestration.findThread` WebSocket RPC, server-side searchable-message projection with revision tracking, and a shared visible-text presentation module that derives searchable text from the rendered transcript rather than raw message source.
> - The server persists a `searchableMessagesRevision` on projection thread rows (migration 48) and returns paginated, revisioned match results with cursors that reopen the matching detail window; a bounded LRU text cache avoids re-deriving visible text on repeated searches.
> - The web client resolves matches against the composed DOM (including open shadow roots in review diffs), positions virtualized timeline rows, loads the target history window on demand, and highlights occurrences via the CSS Highlight API with a data-attribute fallback.
> - Moves markdown, citation, directive, artifact-template, provider-skill, and file-path-display utilities from `client-runtime`/`web` into `packages/shared` so the server and client share one visible-text pipeline.
> - Risk: `OrchestrationThreadShell` gains an optional `searchableMessagesRevision` field and `ThreadDetailPageCursor` gains optional direct-target fields; older cursors and shells without these fields still decode, but any out-of-tree consumers reading raw cursor or shell payloads must handle the new optional keys. Migration 48 adds a non-null column with a zero default to the projection thread table.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 4bd22c3.</sup>
> 
>

### CODE REVIEW

File: packages/shared/src/threadFind.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+function appendMarkdownText(node: MarkdownNode, parts: string[]): void {
+  const block = BLOCK_NODE_TYPES.has(node.type);
+  if (block) parts.push("\n");
+
+  if (node.type === "text" || node.type === "inlineCode" || node.type === "code") {
+    parts.push(node.value ?? "");
+  } else if (node.type === "break") {
+    parts.push("\n");
+  } else if (node.type !== "image" && node.type !== "imageReference" && node.type !== "html") {
+    for (const child of node.children ?? []) appendMarkdownText(child, parts);
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply packages/shared/src/threadFind.ts]: Fixed in 82c3af207. The shared projection now parses Markdown plus raw HTML through a sanitized remark-to-rehype pipeline, so visible HTML text participates in matching while script content is excluded. Added a focused regression test.

File: apps/server/src/orchestration/Layers/ProjectionSnapshotQuery.ts [hunk 2]
```diff
@@ -401,6 +411,23 @@ const makeProjectionSnapshotQuery = Effect.gen(function* () {
...
+    if (cached !== undefined) {
+      threadFindTextCache.delete(key);
+      threadFindTextCache.set(key, cached);
+      return cached;
+    }
+    const visibleText = threadFindVisibleText(row.source, row.text);
+    threadFindTextCache.set(key, visibleText);
+    if (threadFindTextCache.size > THREAD_FIND_TEXT_CACHE_MAX_ENTRIES) {
+      const oldestKey = threadFindTextCache.keys().next().value;
+      if (oldestKey !== undefined) threadFindTextCache.delete(oldestKey);
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/server/src/orchestration/Layers/ProjectionSnapshotQuery.ts]: Fixed in 82c3af207. The cache now stores the complete derived projection per thread revision, bounded by 4 threads and 16M UTF-16 characters, so repeated queries do not churn on threads above 2,000 messages.

File: apps/web/src/components/chat/MessagesTimeline.tsx [hunk 3]
```diff
@@ -1465,24 +1506,31 @@ function AssistantTimelineRow({ row }: { row: Extract<TimelineRow, { kind: "mess
   return (
     <>
       <div className="relative min-w-0 px-1 py-0.5">
-        <AssistantCitationSource
+        <ThreadFindSource
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/web/src/components/chat/MessagesTimeline.tsx:1514]: Fixed in 82c3af207. Server projection now includes sanitized visible raw-HTML text, and code-block header chrome is marked with data-thread-find-exclude so DOM ranges stay aligned. A DOM regression test covers excluded header text.

File: apps/web/src/components/chat/MessagesTimeline.tsx [hunk 4]
```diff
@@ -2268,12 +2316,13 @@ const CollapsibleUserMessageBody = memo(function CollapsibleUserMessageBody(prop
...
   skills: ReadonlyArray<Pick<ServerProviderSkill, "name" | "displayName">>;
   markdownCwd: string | undefined;
+  forceExpanded?: boolean;
   footer?: ReactNode;
 }) {
   const [expanded, setExpanded] = useState(false);
   const hasVisibleBody = props.text.trim().length > 0 || props.terminalContexts.length > 0;
   const canCollapse = hasVisibleBody && shouldCollapseUserMessage(props.text);
-  const isCollapsed = canCollapse && !expanded;
+  const isCollapsed = canCollapse && !expanded && !props.forceExpanded;
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/web/src/components/chat/MessagesTimeline.tsx:2330]: Fixed in 82c3af207. A user message force-expanded for a search match no longer exposes an unusable collapse control; outside that state, aria-expanded and the label derive from the effective collapsed state.

File: apps/server/src/ws.ts [hunk 5]
```diff
@@ -1401,6 +1403,20 @@ const makeWsRpcLayer = (
             ),
             { "rpc.aggregate": "orchestration" },
           ),
+        [ORCHESTRATION_WS_METHODS.findThread]: (input) =>
+          observeRpcEffect(
+            ORCHESTRATION_WS_METHODS.findThread,
+            projectionSnapshotQuery.findThread(input).pipe(
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/server/src/ws.ts:1409]: Fixed in 82c3af207. The find query now filters threads with threads.archived_at IS NULL, with a server regression test confirming archived threads return no results.

File: apps/web/src/index.css [hunk 6]
```diff
@@ -516,6 +516,16 @@ html[data-mobile-composer-route-transition="true"]::view-transition-old(t3-mobil
   ::highlight(t3-assistant-citation-comment) {
     background-color: color-mix(in oklab, var(--primary) 45%, transparent);
   }
+
+  ::highlight(t3-thread-find) {
+    color: inherit;
+    background-color: color-mix(in oklab, var(--primary) 48%, transparent);
+  }
+
+  [data-thread-find-active="true"] {
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/web/src/index.css]: Fixed in 82c3af207. The fallback active-match styling is now expressed as Tailwind data variants on ThreadFindSource; the global CSS fallback rule was removed.

File: apps/web/src/components/chat/ThreadFindBar.tsx [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+  return (
+    <div
+      role="search"
+      aria-label="Find in conversation"
+      className="pointer-events-auto absolute top-2 right-3 left-3 z-30 flex h-9 items-center gap-1 rounded-lg border border-border/70 bg-popover/95 p-1 shadow-lg backdrop-blur-md sm:left-auto"
+      data-thread-find-bar="true"
+    >
+      <SearchIcon className="ml-1 size-3.5 shrink-0 text-muted-foreground" aria-hidden="true" />
+      <input
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/web/src/components/chat/ThreadFindBar.tsx]: Fixed in 82c3af207. ThreadFindBar now uses the existing InputGroup, InputGroupAddon, and InputGroupInput primitives while preserving the compact layout.

File: apps/web/src/components/chat/ThreadFindBar.tsx [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+  useEffect(() => {
+    const closeOnEscape = (event: KeyboardEvent) => {
+      if (event.key !== "Escape") return;
+      event.preventDefault();
+      event.stopPropagation();
+      onClose();
+    };
+    window.addEventListener("keydown", closeOnEscape, true);
+    return () => window.removeEventListener("keydown", closeOnEscape, true);
+  }, [onClose]);
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/web/src/components/chat/ThreadFindBar.tsx:78]: Fixed in 82c3af207. Escape handling is scoped to key events bubbling from the find bar instead of a capturing window listener. Visual verification confirms Escape closes the command palette first and leaves find open.

File: apps/web/src/components/CommandPalette.tsx [hunk 9]
```diff
@@ -1629,6 +1640,20 @@ function OpenCommandPaletteDialog(props: {
...
+      value: "action:find-in-conversation",
+      searchTerms: ["find", "search", "conversation", "thread", "messages"],
+      title: "Find in conversation",
+      icon: <TextSearchIcon className={ITEM_ICON_CLASS} />,
+      shortcutCommand: "thread.find",
+      run: async () => {
+        openThreadFind();
+      },
+    });
+  }
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/web/src/components/CommandPalette.tsx:1658]: Verified as already guarded. On a /draft/$draftId route, useHandleNewThread sets routeThreadRef to null, so activeThread is null. The command palette requires activeThread before adding Find in conversation, so the action is not emitted for drafts. No code change was needed.

**Comment by Lucenx9** [comment]
@coderabbitai review

**Comment by Lucenx9** [comment]
@codex review

File: apps/web/src/components/chat/ThreadFindBar.tsx [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+      <InputGroup
+        variant="ghost"
+        className="h-7 min-w-0 flex-1 border-0 bg-transparent shadow-none hover:bg-transparent has-[input:focus-visible]:bg-transparent has-[input:focus-visible]:ring-0 sm:w-56 sm:flex-none **:[input]:h-7 **:[input]:px-1 **:[input]:text-sm **:[input]:leading-7"
+      >
+        <InputGroupAddon className="ps-1 pe-0">
+          <SearchIcon className="size-3.5 shrink-0 text-muted-foreground" aria-hidden="true" />
+        </InputGroupAddon>
+        <InputGroupInput
+          ref={inputRef}
+          type="search"
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/web/src/components/chat/ThreadFindBar.tsx]: Fixed in the latest branch: the find field now uses the InputGroup ghost variant and InputGroupInput size sm without overriding the shared primitive styling.

File: packages/shared/src/threadFind.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    for (const child of node.children ?? []) appendHtmlText(child, parts);
+  }
+
+  if (block) parts.push("\n");
+}
+
+/** Text users can actually find in rendered message bodies, with layout whitespace collapsed. */
+export function threadFindVisibleText(role: "user" | "assistant", text: string): string {
+  const markdownRoot = parser.parse(threadFindMessageMarkdown(role, text));
+  const root = parser.runSync(markdownRoot) as ThreadFindHtmlNode;
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply packages/shared/src/threadFind.ts]: Fixed: user and assistant messages now use separate parsers. User HTML stays literal, matching parseRawHtml false; assistant HTML still follows rendered output.

File: packages/shared/src/threadFind.ts [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    const match = TRAILING_USER_CONTEXT.exec(prompt);
+    if (!match) break;
+    const kind = match[1] ?? "";
+    const summary = visibleContextSummary(kind, match[2] ?? "");
+    if (summary.length > 0) summaries.unshift({ kind, text: summary });
+    prompt = prompt.slice(0, match.index).replace(/\n+$/u, "");
+  }
+  // Review comments are rendered as cards. Their attributes are metadata, but
+  // the comment and diff between the tags are visible transcript content.
+  const withoutReviewTags = prompt.replace(/<\/?review_comment\b[^>]*>/giu, "");
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply packages/shared/src/threadFind.ts]: Fixed: review cards now project the visible file path, section title, range label, and body in rendered order, with regression coverage.

File: docs/internals/thread-find.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+and zero-based occurrence within that message. Response size therefore stays bounded even when a
+short query occurs thousands of times.
+
+The projection query reads only settled user messages and canonical final assistant messages for
+the selected thread. [`threadFind.ts`][text] converts Markdown to rendered text, condenses
+composer-only context blocks, collapses layout whitespace, and counts case-insensitive,
+non-overlapping matches. No index or backfill is required. If profiling shows that parsing a single
+large thread is material, a derived searchable-text projection can be added behind the same RPC.
+Repeated queries reuse a 2,000-entry in-process LRU keyed by message revision; the bound prevents a
+long-running remote server from retaining every message it has ever searched.
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply docs/internals/thread-find.md]: Fixed: the internals doc now describes the per-thread derived projection cache and its four-thread / 16M UTF-16 character bounds.

File: apps/server/src/orchestration/Layers/ProjectionSnapshotQuery.ts [hunk 14]
```diff
@@ -911,6 +973,46 @@ const makeProjectionSnapshotQuery = Effect.gen(function* () {
...
+          messages.text,
+          messages.updated_at AS "updatedAt"
+        FROM projection_thread_messages AS messages
+        INNER JOIN projection_threads AS threads
+          ON threads.thread_id = messages.thread_id
+        INNER JOIN projection_projects AS projects
+          ON projects.project_id = threads.project_id
+        WHERE messages.thread_id = ${threadId}
+          AND threads.deleted_at IS NULL
+          AND threads.archived_at IS NULL
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply apps/server/src/orchestration/Layers/ProjectionSnapshotQuery.ts:1088]: Not applicable: archiving the currently open thread immediately navigates to a new draft in useThreadActions, and archived threads are only exposed in Settings where the chat view and find command are not mounted. Keeping the server query scoped to active threads matches that lifecycle.

File: packages/shared/src/threadFind.ts [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+import { unified } from "unified";
+
+type ThreadFindHtmlNode = {
+  readonly type: string;
+  readonly value?: string;
+  readonly tagName?: string;
+  readonly children?: ReadonlyArray<ThreadFindHtmlNode>;
+};
+
+const parser = unified()
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply packages/shared/src/threadFind.ts]: Fixed: local file links now project the same visible path or basename label used by chat rendering, with regression tests.

File: packages/shared/src/threadFind.ts [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+  .use(remarkParse)
+  .use(remarkGfm)
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply packages/shared/src/threadFind.ts]: Fixed: the shared projection now applies remarkCodexDirectives and has tests for file citations and artifact templates.

[the remark this answers is not in the document]
Comment (Lucenx9) [reply packages/shared/src/threadFind.ts]: Fixed: the projection now applies remarkGithubAlerts, so alert markers are removed as in rendered chat; regression tests cover it.

File: packages/shared/src/threadFind.ts [hunk 17]
```diff
@@ -0,0 +1,540 @@
...
+    const suffixByPath = new Map<string, string>();
+    for (const fileNode of fileNodes) {
+      fileNode.node.children = [
+        {
+          type: "text",
+          value: threadFindFileLabel(fileNode.position, fileNode.path, suffixByPath),
+        },
+      ];
+    }
+
```
[the remark this answers is not in the document]
Comment (Lucenx9) [reply packages/shared/src/threadFind.ts:256]: Fixed in 4bd22c3f3: the projection now resolves file paths against workspaceRoot with the same shared resolver as ChatMarkdown, carries the Markdown/inline suffix map into the raw-HTML stage, and reuses it for matching HTML links. HTML-only links remain basename-only to match the current UI. Added shared projection and React rendering regressions.
